### ISSUE TITLE
[RFC] Pollution revamp

### ISSUE THREAD

**Issue by TheElan** [GTNewHorizons/GT-New-Horizons-Modpack#8696]
### Your GTNH Discord Username

Elan

### Your Pack Version

2.1.1.0

### Your Proposal

As I've noticed current problem with pollution revamp is a lack of a structured requirements.

First part of my proposal is to gather all ideas on how pollution and it parts should work from a player perspective under this issue. 

Please check anything posted before making a proposal. If you have a more concrete idea on how it should work and it isn't already under this issue - post it here. If you can split your suggestion into parts with clear boundaries. For example in one post you propose on how to add a pollution to the world and in the other on how to remove pollution from the world and so on.

Second part of it is voting under posts with :+1: or :-1:.

After some time have passed we can collect most up-voted/best ratio ideas, filter it using `DreamMaster::doYouLikeThis` predicate, add some more from `DreamMaster::polutionMustDoThis`. Having this set it would be possible to compose some structured requirements for the revamp.

Note: this is a proposal for high-level ideas understandable for you common player without any modding experience. Discussion on how to better organize things from code perspective better wait till we have a complete list of features. After that it can be discussed in separate issue\discord. Same goes for performance discussions.

Note 2: If you found somebody suggest on which you have an improvement - just create your own post, referencing proposal and stating what you would like to change\expand in it.

### Your Goal
Get a narrowed down list of ideas for the revamp which would allow somebody to start working on design and implementation of the revamp.

### Your Vision

This is a list of things pollution should be. Please provide your opinions not related to describing your concept on discord. This will avoid flooding this thread with irrelevant/subjective/optional information.

#### Example proposition 1:

> #### Pollution types
> - Air pollution:  it hard to breath without some kind of protection
> - Toxic pollution: applies long lasting effects (e.g. reduced health) on entities which spend too much time in area with this without protection
> - Radioactive pollution: when not protected you accumulates radiation. After accumulating too much of it you will start experiencing effects radiation poisoning and if it is a lethal does of radiation - you'll basically get count down timer till you just die without treatment.
>
> This way it will be more "realistic" and would also bring more variety into game-play by requiring different strategies for pollution management.

#### Example proposition 2:
> #### Personal protection
> I propose expanding personal pollution protection. Having a hazmat suit as a silver bullet for any level (and type, see my pollution type proposal) is kind of boring.
> This protection should be tired and each tir would help to protect up to certain amount of pollution, after which it is less\not effective at all
> Low tier:
> - wet cotton mask (becomes dry with time and less effective until soaked in water)
> - herb medicine making you immune to some effects (0.5-1.5h)
> Mid tier:
> - one-use breathing filters
> - more effective solutions to wet cotton mask in
> - pills/potions/magic consumables with long lasting effects (at least 4-6)
> - low level magic artifacts forming a "filtering" field around you
> Late tier:
> - Hazmat suit/etc
> - permanent immunity shots
>
> With this you would have more options to solve your pollution problems and even live with it. 

#### Example proposition 3:
> #### Single configuration
> It should be possible to configure pollution in one place (e.g. how much which machine introduces/removes pollution, effect durations, etc)

### Final Checklist

- [X] I have searched this issue tracker and there is nothing similar already. Posting on a closed issue saying *I like this feature please reconsider adding it* will prompt us to investigate and reopen it once we confirm your report.
- [X] I believe there is nothing similar in the pack already, or the existing solution isn't good enough.
- [X] I understand this change request may not attract enough attention and thus not be implemented.
- [X] I understand this change request may be rejected due to other community members thinking it's inappropriate.
- [X] I believe this feature would make the pack better.

**Comment by leagris** [comment]
Why were mission statements removed. Are we running in circle again with ideas on how to make pollution annoying in some way (we have had plenty of it already). Sorting idea by type is still missing a mission statement. 

- Why shall there be a pollution?
- What do we expect pollution to achieve in GTNH?
- Do we want pollution to incentivise player on a preferred gameplay path; and if so, it need to be described without unambiguously?
- Do we want pollution to slow down player progression even more?
- De we want pollution to slow down high level players?

Previous pollution implementation was suspended because of the following reasons:
1. It had no mission statement and every contributor made changes with diverging interpretation and incompatible scopes.
2. It was more taxing on low level players and high level players had ways to ignore it entirely.
3. Its implementation was broken and damaged by multiple changes on rates and scope by persons who were not on the same tune and gauges.
4. Its implementation was bugged and had effect spawns inter-dimensionally.
5. Pollution could be used as griefing on SMP servers.
6. Pollution did not provide content for players, did not help server owner to fairly allot server resources across players, did not favour collaborative game-play, did not improve or incentivise players into saner or better game-play. To sum it up, pollution was landed like an alien without a mission statement and everybody had to deal with the annoyance and perpetual adjustments that tuned out to always be unfair, and damaging gameplay.

**Comment by bombcar** [comment]
Mission statement: prevent afkings like @bombcar from just using the BBF forever. The end.

**Comment by Folanlron** [comment]
As long as I can turn it off on my server/SP world..

It's a game not real life, and is the stupidest idea ever implemented into Minecraft

**Comment by TheElan** [comment]
> As long as I can turn it off on my server/SP world..
> 
> It's a game not real life, and is the stupidest idea ever implemented into Minecraft

Rephrasing: Pollution should be optional and easy to remove/disable (e.g removing mod).

**Comment by TheElan** [comment]
#### Source of pollution and way to mitigate it
Converting most of current byproducts into some form of pollution or slug substance and lowering processing cost. You can choose to spend additional energy and effort setup a system which recovers original byproducts out of env pollution substance.

Low-tech machines would just release this into "atmosphere", while more advanced ones will be able to capture it in solid or liquid form to simplify processing.

Trying to void "dirty" byproduct (e.g. deconstructing pipe with liquid, throwing into fire/lava/trashcan) should cause it to be released into "atmosphere".

This will not only provide an incentive for player on cleaning their mess, but also provide them with substantial reward, while respecting "preservation of mass" so it's not a cheesy way to get infinite resources.

**Comment by TheElan** [comment]
It should be possible to configure it as a purely cosmetic thing (no positive\negative side effects).

**Comment by TheElan** [comment]
Not punish player for voiding pollution (e.g. if it's a fluid or a solid) - they just will lose the ability to recover valuables from it

**Comment by bombcar** [comment]
#8883 may be considered too

**Comment by bombcar** [comment]
https://github.com/GTNewHorizons/GT5-Unofficial/pull/752 moves all (or most) existing pollution sources into a config, along with additional options for on/off on things like fog.

**Comment by PolkageistHS** [comment]
So pollution is re-enabled on 2.1.2.0pre2, with default values for LV combustion and gas generators at 200 pollution/sec.  That makes a base all but unusable very quickly, especially since the new LV scrubber is gated behind MV (small turbines need MV assembler, and the GT++ ones don't work).

**Comment by boubou19** [comment]
values didn't change, i just made the configs without any balance decision in mind

**Comment by PolkageistHS** [comment]
Something had to have changed, because before they were 40 LV/80 MV/160 HV, now they're 200 LV/180 MV/160 HV.  Gas turbines used to be 20 LV/40 MV/80 HV, now they're 200 LV/180MV/160 HV.

**Comment by boubou19** [comment]
yeah, dunno what i was doing, good catch

**Comment by MrJakobLaich** [comment]
FIRST OFF, I gotta dump everything I just wrote in discord here, I'll make seperate issues and link them here (I hope I have enough mental energy for all that)
Dumping everything here so it's all in one place and can be found, also there's a massive all of text already that's split into messages, I'll get to reading all that too in a moment, gimme time (feel free to still be angry that I posted all this)

Regarding QOTD ("Questions of the day: Why do we need pollution?  What good does it do?")
Pollution feels cool because it's more realistic than "hehe just power the machine and don't worry"
But right now it's just filling a chunk and giving you some annoying side effects (pretty fast from my experience too)
Also, the hazmat suit has way too little durability D:

So a couple of ideas:
1) Make pollution appear slower, the pack goes on for a long time, no need to fill everything with pollution within days
2) Add 1 more hazmat suits so you don't have to repair the low tech hazmat suit 800 times because you fell 4 blocks and it breaks your boots or because you touched a steam pipe and it completely melts away your whole suit, you just make the better one with rubber / polyethylene in MV and in HV upgrade to nano anyways

Also, repairing the hazmat suit is currently more annoying than just making a new one, kinda weird

So, you have the refined glue suit, low durability but better than nothing

Advanced hazmat suit makes sense but is kind of meh to make (annoying recipe for something that everyone's basically forced this much to use) and also can't be repaired AFAIK, needs to use an MV-circuit instead of an HV one, then the recipes would be okay
Give it a bit more protection too, it's advanced afterall

Polyethylene end-of-MV power hazmat suit with good durability and offering even a bit of protection (little less than iron would be okay imo, still bad compared to everything at that tier but better than the other hazmat suits)

REASON for this:
You have so much pollution that you are forced to wear the hazmat suits anyways so it's either extremely annoying because you have to repair them every 3 seconds (currently) and we buff the suit (or add higher tier suits) or we nerf the pollution, which is what I'll talk about next (remember, buff is most of the time better than nerf xD)

Asked Alastor's Game (Fuel Force) (he's the magic person when I have any magic questions xD)
And thaumic restorer could repair this stuff much easier than any crafting recipe so I'd say we remove the hazmat suit repair recipe and give the refined glue suit a bit more durability
Restorer is MV, you just need ender eyes which can be traded for, but I'd prefer it if the ender eye recipe got some thaumcraft equivalent so people don't instantly see that you can get them in MV, normal recipe is HV afterall... (why?), if we keep the HV recipe nobody would ever even notice that the thaumcraft one even exists lol

You get poly and then mid-end MV you use it to make some better hazmat suit and stay in your new skin until you get the nano suit

3) most of the effects that pollution give you currently are there to make you stop polluting or to put on your hazmat suit or start air filtering, see above what I think about them lol
I currently have one chunk with 300k+ gibbl and going up at about 40 per second
If the wiki is still correct then I had at least 750k at some point, which is a lot considering that that wasn't that long since 2.1.2 came out
So, imo the pollution should go up a bit slower than it is right now, at least for LV machines
(Yes, I know about air scrubbers but that's not the point right now, I'll maybe get to that too)
Do steam BBF / coke ovens produce pollution??

The effect tiers should be changed, you should notice a change before you get f'ed up by it
So, different tiers
The first ones for noticing, shouldn't affect gameplay too much
The second one for actually hampering gameplay, should make you seriously consider putting on the (mandatory...) hazmat suit, you still have some time before you can't anymore
The third for breaking you if you haven't put on the suit
The fourth "you die lol" when you somehow made it through the last tier without having a hazmat suit on

Every tier should also add a randomly selected number of effects from (all) lower tiers of effects too
Also, the lower the tier compared to the current pollution effect tier the higher the effect duration and effect level
So, tier 3 could have 1 tier 3 effect and any tier 2 and any tier 1 effect too, good luck with that lol
So, for example, t3 pollution levels could have poison 1 for 23s, slowness 2 for 40s, mining fatigue 2 for 35s, hunger 3 for 40s and weakness 3 for 30s

T1 weakness (10-20s), hunger (5-10s), nausea (3s)
T2 slowness (10-30s), mining fatigue (20s)
T3 blindness (20-40s), poison (20-30s)
T4 wither (random duration 40-120s, yeah 120 for people with many hearts :troll:, as I said, good luck with this)
I'd add more effects to this but idk what Volatility or Ill Fitting do so I can't really use them for this :shrug:
Also, I'd add some warp effects too, like the one that makes the screen completely dark instead of just the blindness debuff, then I could add more tiers for pollution, which would be nice

Before I do the vegetation and world effects, I'll list some warp effects that would make a bit of sense from what they sound like, maybe we could add some to the pollution stuff

Insomnia, but make it say "You cough yourself awake" or something like that, probably T2 or T3
Thunderstorm, you ruined the environment T3 or T4
Whatever gives you the completely dark screen warp effect, put that into T3 instead of blindness and put blindness (10-20s) into T2

Huh, that's it lol, there's not much more there that looks good imo

Regarding gadgets, GC slots and stuff like that
We could add some air tank you have to use, afterall you can't really breath inside a plastic bag, so why should you be able to inside a hazmat suit without air tanks
We could add some low tech breathing equipment and stuff like that, different tiers of course, who doesn't like upgrading their stuff from time to time
So, probably 4 tiers of breathing stuff until you can make the nano suit
Put the hazmat suit on and it protects you from pollution stuff
But when you put on the hazmat suit you slowly lose oxygen, same as when you are under water
Put the breathing stuff into the gadget slot and you don't lose air
Air capacity depends on breathing stuff tier, first tier works for 1h, second for 3h, third tier for 8h and fourth tier for 12h (anything less is 
 bad because just the suit is already annoying enough)
Refill them by crafting together with an (or multiple in case of higher tier ones) oxygen cell, shouldn't consume the empty cell

I still don't have anything in the slots except for the holo glasses so the GC, witching gadgets from Alastor and tinkers should be fine, probably should use the gadgets or GC slots, probably gadgets ones, I'd not use tinkers because you maybe make the extra hearts things at some point so either GC or WG, I don't think both matter because when you make the GC space suit it should automatically do the oxygen thing anyways (same with the nano suit and everything else later on) so once you have that all you don't need it in the GC slot anyways

So, putting stuff into the GalacticCraft slots should work until HV when you have to make the space suit, the GC slots aren't used until then and after that you shouldn't need the hazmat suit anymore because you have either the GC space suit or the nanosuit

Bombcar:
"The real deal is pollution isn’t fun"
This is exactly what I'm trying to fix, the main issues I see are:
1) Pollution right now is only barebones, not really much to it except "let's bully people into using the hazmat suit and air scrubbers"
2) hazmat suit sucks, way too low durability so you have to repair it every 3 seconds, see top part of this message
3) yeah, it's annoying and that's it, only negative stuff that doesn't have much reason to it

Alright, vegetation and whatever's gonna happen to your blocks
I'll only write down what additional stuff we could do or what we could change
I think most of the vegetation stuff is alright the way it is, just a few tweaks
All flowers should turn into dead bushes, idk why they would drop as an item, if you plant stuff in pollution then that's on you anyways, that's why we have T1 and T2 as warnings, if you still have plants when you get to T2-T3 levels it's your fault

Quick input regarding when what tier happens because I'd forget otherwise:
(remember, pollution would happen a bit slower than currently, so 100k should happen a lot slower than instantly)
T1 happens at 100k gibbl, that's the warning, so no big debuffs
T2 happens at 300k, this is where you actually get annoyed and bullied by the game so you start using the hazmat constantly
T3 happens at 750k, vegetation destruction happens at 600k, plants don't like bad air... (dead plants would also be a good warning for you if you somehow still haven't made a hazmat suit)
T4 is at 1M gibbl, for now there's no tier above that
T5 is probably around 1.25M - 1.5M gibbl
Sour rain is at 900k gibbl
Late game you either have air filters running or just wear your suit 24/7
Alright, back to plants

All flowers should turn into dead bushes
Cropsticks should become the "Dead Grass" item from "Biomes O' Plenty" and drop the cropstick
If we want to break people we just switch the plant on the cropstick to weed, that'd be !FUN!, good luck if you don't notice that (you should, after all you get enough warnings)

Alright, I don't have any plant ideas anymore, I'll add more when I figure something out

Oh, yeah, also:

When a tree grows it should remove some amount of pollution, when any vegetation thing happens it should remove some pollution
So, if a flower gets turned into a bush it removes, made up number for now, 1k gibbl
So if you plant enough trees early on it should reduce pollution by a  bit, lowest tech air scrubber
ANY pollution effect should reduce pollution by some amount, afterall it some reaction which uses pollution as a component
It just shouldn't be an insane amount, just enough to reduce it by a bit if you have a tree farm for example

On we go to biomes

Biomes, I do remember someone talking about how pollution could change a biome into some wasteland and I really like the idea
In T5 (gotta figure some stuff for this at some point) we could change water into wastewater and lava into stone, that'd be cool maybe

Making water become wastewater and same with lava would break rock breakers

Also, what I noticed is chunkwise pollution is a little bit small, maybe making it 8x8 chunks would be cooler, that way you don't have one 500k chunk and one 13k chunk next to each other, also make a generator 8² chunk only leak into the 8 8² chunks next to it and not any further, that way you get a 192x192 block of polluted chunks
God, I really hope putting pollution into ungenerated chunks doesn't break the game LMAO
If it would then, yeah, idk, having one chunk with 500k and one with 13k makes no sense, so maybe at least make it be 64x64 blocks instead of one chunk that stores pollution

Pollution above ground should spread / go away way faster, because there's a LOT of space outside for stuff to disperse to

Also, the main idea here:
Pollution is everywhere in a chunk, even if I make a hole in the ground, put the machine inside and seal it, somehow the pollution is above ground D:
Pollution inside a mountain should stay down there, that way you have a reason for the air scrubbers and all that, maybe even air vents and things like that
I do realize that calculating some A* route to the outside every few ticks is going to tank TPS by a massive amount but we could
1) spread it out over multiple ticks (we're kind of doing something like that for pollution already in the pollution controller)
2) make the route grid bigger, so instead of every block we only check every second block or some variation of this, but that could become a problem so I'd not recommend this at all
3) We take a faster but less precise A* algorithm? (has been about a year since I've looked into this stuff so I gotta read about that again)
4) idk, give me some better ideas
We apparently just call if(block.getWorld().getHighestBlockYAt(block.getLocation()) == block.getBlockY()) for every block we come across while we check stuff (if there's a better way to check if any random block is "outside" then please tell me, because this is the best thing I could come up with :p)
We could also limit the size of whatever cave we check to 100x100x40 blocks or something, at if we go above that we assume it's outside or something like that
That way you could make a "cleanroom" for your plants (probably outside) and a "dirtyroom" with your machines (and a real cleanroom inside that lol)


DISCORD discussion I had with bombcar after all this stuff: (I cut out all other discussions because they didn't really add anything that new to what I've had alreaady written up there)
bombcar — heute um 04:56 Uhr
The chunks now store pollution “underground” if you will
So I think it can spread without chunk loads
Same way oil works
Also 500k next to 13k should balance out
Dissipation may need tuning if that persists
bombcar — heute um 04:59 Uhr
You’re trying to solve the same problem
As the oxygen on space station
MrJakobLaich (HV) — heute um 04:59 Uhr
Then I'm really for 8x8 chunk blocks for storing pollution and spreading to all 8 blocks around it, but not further
Instead of using normal chunks, because this way most of your base would have kind of the same pollution level and not like it is now, have 300k here and not really any pollution 3 trees over in the next chunks
bombcar — heute um 04:59 Uhr
And likely the “flooding” code from hydrodam is the way to do it
MrJakobLaich (HV) — heute um 05:00 Uhr
Oh yeah that's a thing, I've never really used much of GC so I didn't remember xD
MrJakobLaich (HV) — heute um 05:00 Uhr
That sounds promising
bombcar — heute um 05:00 Uhr
Problem with anything like that is you can abuse the mechanics
And separate by a chunk to get perfectly clean
MrJakobLaich (HV) — heute um 05:01 Uhr
What do you mean?
bombcar — heute um 05:01 Uhr
If you have a hell chunk
And it will ONLy spred to the 8 around it
And then stop
You move two chunks away and you’re fine
So you chunkload and remote control and ignore he pollution
MrJakobLaich (HV) — heute um 05:02 Uhr
Yeah but that's why I'd make it not chunk based but 8-chunks by 8-chunks
So 128x128 blocks
So you have an area of 384x 384 where it should equalize the pollution
24x24 chunks total
bombcar — heute um 05:03 Uhr
Yeah that’s closer to what we should have
But I don’t know the spread rate
Too slow and you get pollutiontowers
Too fast and it dissipate who cares
MrJakobLaich (HV) — heute um 05:04 Uhr
Also, this way you could actually see pollution ruining your forests and stuff and it would take all machines into account
MrJakobLaich (HV) — heute um 05:04 Uhr
I'd play around and at some point it's gonna be good enough
bombcar — heute um 05:04 Uhr
I think this
Could be interesting
MrJakobLaich (HV) — heute um 05:05 Uhr
That's why I said, it's not gonna dissipate into anything further than the next "chunkblock"
bombcar — heute um 05:05 Uhr
Make Polly debuffs not hit you for like 10m
MrJakobLaich (HV) — heute um 05:05 Uhr
So it's contained and if you produce too much it's still gonna fill the center block
bombcar — heute um 05:05 Uhr
Then build at chunk block borders
Like@old sims Oct
MrJakobLaich (HV) — heute um 05:05 Uhr
polly?
oh
bombcar — heute um 05:05 Uhr
Old SimCity
MrJakobLaich (HV) — heute um 05:05 Uhr
pollution
bombcar — heute um 05:05 Uhr
Pollution
Here’s my idea
MrJakobLaich (HV) — heute um 05:05 Uhr
That's not gonna do much because it's a 128x128 border
You'd have to make your base MASSIVE
bombcar — heute um 05:06 Uhr
If Polly doesn’t hit you until 10m@pr so
You have a reason to build railroads
MrJakobLaich (HV) — heute um 05:06 Uhr
Like, so massive it's useless in MV
bombcar — heute um 05:06 Uhr
To your bbfs
MrJakobLaich (HV) — heute um 05:06 Uhr
Which is cool
MrJakobLaich (HV) — heute um 05:06 Uhr
oh yeah true but they'd have to be way less expensive
You'd need an insane amount of stuff for that
Also, we can change pollution limits, that's why I said, you should first notice small debuffs before it hits you with wither and stuff like that
So you have time to adapt and fix stuff
because otherwise your plants just vanish over a night
In server chunkloading cases
I'll add all this to the github message
Good convo stuff
bombcar — heute um 05:11 Uhr
what we need really
are userstories


Anyways, these are all my ideas regarding pollution and what I think would be cool, if you like or dislike something, please tell me what I could improve about it :)

**Comment by mmcmain** [comment]
Balance consideration (From a player that likes the pollution mechanic):

The multiblock is nice, but it can't keep up with 2 MV blast furnaces with Kanthal coils unless you also give it Absorbtion filters, which is great. However Centrifuging an Absorbtion filter at MV is only an 80% chance of returning the frame. This is overly punishing at MV considering how hard it is to make the filter frame. By the time you get to EV and can get the 100% return it is no longer a pain point for you.

I see a couple decent alternatives to balance it:

1. Make the centrifuge always return the frame (remove the chance of loss).
2. Add two new frame recipes: Aluminium and Titanium. The first has an 60% chance of loss, stainless 80% and Titanium has no chance of loss.

An LV Multiblock air purifier with absorption filters will slow the progress of 2 MV blast furnaces enough that you can run them for a reasonably long time (maybe 30 mins?) and be right on the edge of debilitating pollution. That part feels really well balanced to me as an MV player. If I had an Aluminium version of the frame available I would keep it full with a stack of filters knowing it is an Aluminum sink.

**Comment by Gateswong** [comment]
As of the posting date, I just figured out the pullotion seems causing significant LAG. I observed my server lags for 2 ~ 5 seconds (having below 0.1 tps) around every 40 seconds. After some profiling, I found that the lag spike happens exactly a thread is happening, which invokes the following function:

https://github.com/GTNewHorizons/GT5-Unofficial/blob/cfe0df76ab423d4dc9853d3ef5625ec2983f3523/src/main/java/gregtech/api/util/GT_ChunkAssociatedData.java#L185

And it has 2 implementations, pollution and underground fluid. Based on how pollution works, I suspect it might be major culprit to this problem.
